TensorFlow Transformer Part-2

phi-dbq opened the pull request:
Utilities to infer tensor and operation names: `as_tensor_name` and `as_op_name`.
They implement the following functionalities.
`{tf.Tensor, tf.Operation, tf.Tensor name, tf.Operation name} |-> name of a tf.Operation`
`{tf.Tensor, tf.Operation, tf.Tensor name, tf.Operation name} |-> name of a tf.Tensor`

These two methods differ from existing ones in that if the input is already a string type, they will simply go and infer the name of the graph element (either tensor or operation). 
These are useful when working outside the context of a `tf.Graph` object. 
Existing methods such: `tensor_name`, `op_name`, `get_tensor` and `get_op` all require a `tf.Graph` object as input.

phi-dbq committed:
update utils

phi-dbq committed:
tests

sueann commented on python/tests/tests.py:
```diff
@@ -29,6 +29,9 @@
 from pyspark.sql import SparkSession
 
+class PythonUnitTestCase(unittest.TestCase):
+    # Just the plain test unittest.TestCase, but won't have to do import check
```
do you mean the unittest2 check above? if so make that explicit.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,49 @@
+        tnsr = tf.constant(1427.08, name=op_name)
+        graph = tnsr.graph
+
+        self.assertEqual(op_name, tfx.as_op_name(tnsr))
```
what's the difference between as_op_name and op_name ?

phi-dbq replied:
`op_name` takes a graph and a `tfobj_or_name`  argument, and check if the latter is part of the former. 
`as_op_name` returns a valid tensor name if the input is a string. Otherwise, it uses the graph of the `tfobj` and call `op_name`. 
Thus `op_name` is a bit more strict than `as_op_name`.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,49 @@
+from ..tests import PythonUnitTestCase
+
+class TFeXtensionGraphUtilsTest(PythonUnitTestCase):
+
```
each test function should test a particular aspect of a particular function in the tfx module, so it's easy to understand what exactly is being tested, and thus if a test fails it's immediately obvious what exactly failed.  could you restructure so that is true? e.g. a few I see embedded in the existing tests below are:
```
def test_op_name():
  # test that op_name from an op name equals input

def test_op_name_from_tensor():
  # test that op_name from a tensor is correct

...
```
If there is shared set-up code, you can make helper functions for them.

phi-dbq replied:
pending on the result on test refactoring in PR-1

sueann reviewed (changes requested):
Could you add a description to the PR so I can understand why this change is needed?

phi-dbq committed:
fix style

Using the following YAPF style
========================================================
based_on_style = pep8
ALIGN_CLOSING_BRACKET_WITH_VISUAL_INDENT=True
BLANK_LINE_BEFORE_NESTED_CLASS_OR_DEF=False
COLUMN_LIMIT=100
SPACE_BETWEEN_ENDING_COMMA_AND_CLOSING_BRACKET=False
SPLIT_ARGUMENTS_WHEN_COMMA_TERMINATED=True
SPLIT_BEFORE_FIRST_ARGUMENT=False
SPLIT_BEFORE_NAMED_ASSIGNS=False
SPLIT_PENALTY_AFTER_OPENING_BRACKET=30
USE_TABS=False
========================================================

phi-dbq committed:
Merge branch 'tf-transformer-part1' into tf-transformer-part2

phi-dbq committed:
Merge branch 'tf-transformer-part1' into tf-transformer-part2

phi-dbq committed:
Merge branch 'tf-transformer-part1' into tf-transformer-part2

sueann commented:
Can you say in the PR description which PR this is used in? Trying to understand how it is used (outside of tests).

sueann reviewed (changes requested):
i think i understand op_name and as_op_name better now. will we break a lot of things if we changed this function to be def op_name(tfobj_or_name, graph=None) and rolled in as_op_name into it? Same for tensor_name.

then we have only one function, with the assumption that if the graph is given, we'll verify that the tensor / op name is legit.

having all these functions that are almost the same is confusing for the user.

phi-dbq committed:
Merge branch 'tf-transformer-part1' into tf-transformer-part2

phi-dbq committed:
refactoring tfx API

phi-dbq reviewed (commented):
Consolidated a bunch of `(as_|get_)?(tensor|op)(_name)?` functions as per the comments.

phi-dbq committed:
test refactoring

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
     """
     Derive tf.Tensor name from an op/tensor name.
-    We do not check if the tensor exist (as no graph parameter is passed in).
+    If the input is a name, we do not check if the tensor exist
```
"If the input is a TensorFlow object, or the graph is given, we also check that the tensor exists in the associated graph."

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
-def as_tensor_name(name):
+def tensor_name(tfobj_or_name, graph=None):
     """
     Derive tf.Tensor name from an op/tensor name.
```
Derive tf.Tensor name from a op/tensor name or object.

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
     """
-    graph = validated_graph(graph)
-    return get_op(graph, tfobj_or_name).name
+    # If `graph` is provided, directly get the graph operation
```
no need for comment

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
+    # If `graph` is provided, directly get the graph operation
+    if graph is not None:
+        return get_tensor(tfobj_or_name, graph).name
+    # If `graph` is absent, check if other cases
```
no need for comment

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
+def op_name(tfobj_or_name, graph=None):
     """
-    Get the name of a tf.Tensor
+    Derive tf.Operation name from an op/tensor name.
```
"name or object."

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
     """
-    Get the name of a tf.Tensor
+    Derive tf.Operation name from an op/tensor name.
+    If the input is a name, we do not check if the operation exist
```
"If the input is a TensorFlow object, or the graph is given, we also check that the tensor exists in the associated graph."

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
-
-    :param graph: tf.Graph, a TensorFlow Graph object
     :param tfobj_or_name: either a tf.Tensor, tf.Operation or a name to either
     """
```
:param graph: should go here?

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
 
-    :param graph: tf.Graph, a TensorFlow Graph object
     :param tfobj_or_name: either a tf.Tensor, tf.Operation or a name to either
     """
```
:param graph: should go here?

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
     """
-    graph = validated_graph(graph)
-    return get_tensor(graph, tfobj_or_name).name
+    # If `graph` is provided, directly get the graph operation
```
no need for comment

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
+    # If `graph` is provided, directly get the graph operation
+    if graph is not None:
+        return get_op(tfobj_or_name, graph).name
+    # If `graph` is absent, check if other cases
```
no need for comment

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
+        return get_op(tfobj_or_name, graph).name
+    # If `graph` is absent, check if other cases
+    if isinstance(tfobj_or_name, six.string_types):
+        # If input is a string, assume it is a name and infer the corresponding operation name.
```
no need for comment

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -89,77 +87,81 @@ def get_tensor(graph, tfobj_or_name):
+        return get_tensor(tfobj_or_name, graph).name
+    # If `graph` is absent, check if other cases
+    if isinstance(tfobj_or_name, six.string_types):
+        # If input is a string, assume it is a name and infer the corresponding tensor name.
```
no need for comment

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+    for tnsr_idx in range(17):
+        tnsr_name = '{}:{}'.format(op_name, tnsr_idx)
+        yield TestCase(data=(op_name, tfx.op_name(tnsr_name)),
+                       description='must get the same op name from its tensor')
```
i'd say 'test tensor name to op name'

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+        yield TestCase(data=(op_name, tfx.op_name(tnsr_name)),
+                       description='must get the same op name from its tensor')
+        yield TestCase(data=(tnsr_name, tfx.tensor_name(tnsr_name)),
+                       description='must get the tensor name from its operation')
```
i'd say 'test tensor name to tensor name'

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+
+def _gen_graph_elems_names():
+    op_name = 'someOp'
+    for tnsr_idx in range(17):
```
do we nee do to test 17?? something like [0, 3, 17] seems good enough. such a list shows more deliberate test cases.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+TestCase = namedtuple('TestCase', ['data', 'description'])
+
+
+def _gen_graph_elems_names():
```
these functions are producing test cases. naming them so is more clear, e.g. `_gen_tensor_op_string_input_tests`

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+                       description='must get the tensor name from its operation')
+
+
+def _gen_wrong_graph_elems_types():
```
nit: _gen_wrong_graph_elem_type_tests

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+            tfx.op_name(data, msg=description)
+
+    @parameterized.expand(_gen_wrong_graph_elems_types)
+    def test_wrong_op_types(self, data, description):
```
is this meant to be `test_wrong_tensor_type(...)` ?

phi-dbq replied:
I made changes to these. The tests are more refined  now.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+    def test_wrong_op_types(self, data, description):
+        """ Must fail when provided wrong types """
+        with self.assertRaises(TypeError):
+            tfx.op_name(data, msg=description)
```
tensor_name ?

phi-dbq replied:
I made changes to these. The tests are more refined now.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+    def test_wrong_op_types(self, data, description):
+        """ Must fail when provided wrong types """
+        with self.assertRaises(TypeError):
+            tfx.op_name(data, msg=description)
```
add to description: " for op_name" to make it clear since these test cases are being used for testing multiple functions.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+
+class TFeXtensionGraphUtilsTest(PythonUnitTestCase):
+    @parameterized.expand(_gen_graph_elems_names)
+    def test_valid_graph_element_names(self, data, description):
```
`test_valid_tensor_op_string_inputs` is more clear.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+    yield TestCase(data=(tnsr_name, tfx.tensor_name(tnsr, graph)),
+                   description='get tensor name from tensor (with graph)')
+    yield TestCase(data=(tnsr, tfx.get_tensor(tnsr, graph)),
+                   description='get tensor from the same tensor (with graph)')
```
the four cases (this and below) require `graph` as input, so we don't need to add "(with graph)" in the description. having it makes it seem like there should be "(no graph)" versions.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+    tnsr = tf.constant(1427.08, name=op_name)
+    graph = tnsr.graph
+
+    yield TestCase(data=(op_name, tfx.op_name(tnsr)),
```
since there are so many test cases here, let's add comments in between sections, e.g. 
```
# Tests for op_name
yield ..
...
# Tests for tensor_name
...
# Tests for get_op
...
# Tests for get_tensor
...
# ...
```

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+                   description='get tensor from the same tensor (with graph)')
+    yield TestCase(data=(tnsr.op, tfx.get_op(tnsr, graph)),
+                   description='get op from tensor (with graph)')
+    yield TestCase(data=(graph, tfx.get_op(tnsr, graph).graph),
```
why would we be worried about this not being true? is there some complex logic in `get_op` we can mess up?

phi-dbq replied:
One basic contract for the `get_op` and `get_tensor` functions is that the returned tensor or operation should belong to the graph that we pass as the second argument.
Otherwise, there could be subtle errors that are hard to debug. For example, we might feed data to a tensor/placeholder on the wrong graph.

sueann replied:
That's a statement about handling bad input. It'd make more sense to test that it raises error if the graph and the op are not compatible, no?

sueann replied:
That sounds like a statement about handling bad input. We should also test that these raise errors if the graph and the op are not compatible.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,92 @@
+
+    yield TestCase(data=(op_name, tfx.op_name(tnsr)),
+                   description='get op name from tensor (no graph)')
+    yield TestCase(data=(op_name, tfx.op_name(tnsr, graph)),
```
we should have tests for getting op_name and tesor_name from op objects as well

sueann reviewed (changes requested):
looks pretty good. thanks for consolidating the op_name tensor_name functions! mostly small comments below.

phi-dbq committed:
PR comments

1. docs in graph/utils.py

phi-dbq committed:
(wip) utils test

phi-dbq committed:
a few more tests for utils

phi-dbq committed:
test update cont'd

phi-dbq commented:
@sueann I updated the PR regarding the last round of comments.

phi-dbq committed:
Merge branch 'tf-transformer-part1' into tf-transformer-part2

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -27,66 +27,135 @@
 
 
-def _gen_graph_elems():
+def _gen_valid_tensor_op_objects():
```
now it seems to be testing all input combos. if so, rename to `_gen_valid_tensor_op_input_combos`

sueann reviewed (commented):
just a few comments

phi-dbq committed:
PR comments

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -71,8 +72,7 @@ def get_op(tfobj_or_name, graph):
     op = graph.get_operation_by_name(_op_name)
-    assert op is not None, \
-        'cannot locate op {} in current graph'.format(_op_name)
+    assert isinstance(op, tf.Operation), 'expect tf.Operation, but got {}'.format(type(op))
```
I think the previous error message was more clear. the current one sounds a little bit like you expected the user to give us the operation object, not our internal invocation of `get_operation_by_name`.

sueann commented on python/sparkdl/graph/utils.py:
```diff
@@ -213,3 +213,10 @@ def strip_and_freeze_until(fetches, graph, sess=None, return_graph=False):
+
+
+def _assert_same_graph(tfobj, graph):
+    if graph is None or not hasattr(tfobj, 'graph'):
```
simpler (no need for the extra return statement):
```
if graph is not None and hasattr(tfobj, 'graph'):
  assert ...
```

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -42,7 +42,31 @@ def _gen_invalid_tensor_or_op_input_with_wrong_types():
+
+    # Test get_tensor and get_op returns tensor or op contained in the same graph
+    yield TestCase(data=lambda: tfx.get_op(tnsr, other_graph),
+                   description='test graph from getting op fron tensor')
```
'test get_op from tensor with wrong graph' is more clear (similarly for other testcases in this method)

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -42,7 +42,31 @@ def _gen_invalid_tensor_or_op_input_with_wrong_types():
+    other_graph = tf.Graph()
+    op_name = tnsr.op.name
+
+    # Test get_tensor and get_op returns tensor or op contained in the same graph
```
'Test get_tensor and get_op with non-associated tensor/op and graph inputs'

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -0,0 +1,191 @@
+    yield TestCase(data=(tnsr.op, tfx.get_op(op_name, graph)),
+                   description='test op from op name')
+
+    # Test get_tensor and get_op returns tensor or op contained in the same graph
```
these tests don't seem necessary now that we have the asserts you've most recently added (we invoke the same function calls in other tests and they will fail if the asserts fail). let's remove them to save on testing and reading time. i like this new way of having the asserts in the methods much better than the tests here, since what we're checking here are really implementation details.

sueann reviewed (changes requested):
almost there! just a few nits. thanks!

phi-dbq committed:
PR comments

phi-dbq commented on python/tests/graph/test_utils.py:
```diff
@@ -128,24 +129,6 @@ def _gen_valid_tensor_op_input_combos():
     yield TestCase(data=(tnsr.op, tfx.get_op(op_name, graph)),
                    description='test op from op name')
 
-    # Test get_tensor and get_op returns tensor or op contained in the same graph
```
Removed these tests as they are covered by other test cases.

phi-dbq commented on python/tests/graph/test_utils.py:
```diff
@@ -47,23 +47,24 @@ def _gen_invalid_tensor_or_op_with_graph_pairing():
 
-    # Test get_tensor and get_op returns tensor or op contained in the same graph
+    # Test get_tensor and get_op with non-associated tensor/op and graph inputs
+    _comm_suffix = ' with non-associated tensor/op and graph inputs'
```
Renamed the description to clarify the intention of each test case.

sueann replied:
' with wrong graph' seems good enough here and more readable.

sueann replied:
just to be clear, my comment was for `_comm_suffix`. the long way to say it for the comment is good since there are now two ways the same thing is described to the user - so it should decrease uncertainty if one of them doesn't quite make sense to the reader.

phi-dbq commented on python/sparkdl/graph/utils.py:
```diff
@@ -216,7 +218,6 @@ def strip_and_freeze_until(fetches, graph, sess=None, return_graph=False):
 
 
 def _assert_same_graph(tfobj, graph):
-    if graph is None or not hasattr(tfobj, 'graph'):
```
Removed the first return statement.

phi-dbq commented on python/sparkdl/graph/utils.py:
```diff
@@ -72,12 +72,13 @@ def get_op(tfobj_or_name, graph):
     op = graph.get_operation_by_name(_op_name)
-    assert isinstance(op, tf.Operation), 'expect tf.Operation, but got {}'.format(type(op))
+    err_msg = 'cannot locate op {} in the current graph, got [type {}] {}'
+    assert isinstance(op, tf.Operation), err_msg.format(_op_name, type(op), op)
```
Changed error message to better align to the intended usage of the function rather than focusing locally on this assertion statement.

phi-dbq reviewed (commented):
Updated regarding latest PR comments.

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -47,23 +47,24 @@ def _gen_invalid_tensor_or_op_with_graph_pairing():
+    _comm_suffix = ' with non-associated tensor/op and graph inputs'
     yield TestCase(data=lambda: tfx.get_op(tnsr, other_graph),
-                   description='test graph from getting op fron tensor')
+                   description='test get_op with from tensor' + _comm_suffix)
```
remove "with"

sueann commented on python/tests/graph/test_utils.py:
```diff
@@ -47,23 +47,24 @@ def _gen_invalid_tensor_or_op_with_graph_pairing():
+                   description='test get_tensor from tensor' + _comm_suffix)
     yield TestCase(data=lambda: tfx.get_op(tnsr.name, other_graph),
-                   description='test graph from getting op fron tensor name')
+                   description='test get_op fron tensor name' + _comm_suffix)
```
fron -> from

sueann reviewed (approved):
lgtm. a few nits for you to fix, but will approve (no further review required).

phi-dbq committed:
PR comments

phi-dbq committed:
TensorFlow Transformer Part-3 (#10)

* intro: TFInputGraph

* tests

* Merge branch 'tf-transformer-part1' into tf-transformer-part3

* and so there is no helper classes

* and into more pieces

* class & docs

* update docs

* refactoring tfx API

* update tfx utils usage

* one way to build these tests

* tests refactored

* test cases in a single class

THis will make things easier when we want to extend other base class functions.

* shuffle things around

Signed-off-by: Philip Yang <philip.yang@databricks.com>

* docs mostly

* yapf'd

* consolidate tempdir creation

* (wip) PR comments

* more tests

* change test generator module name

* TFTransformer Part-3 Test Refactor (#14)

* profiling

* tests

* renamed test

* removed original tests

* removed the profiler utils

* fixes indents

* imports

* added some tests

* added test

* fix test

* one more test

* PR comments

* TensorFlow Transformer Part-4 (#11)

* flat param API impl

* support input graph scenarios

* (WIP) new interface implementation

* docs and cleanup

* using tensorflow API instead of our utilities

* automatic type conversion

* cleanup

* PR comments

1. Move `InputGraph` to its module.

* (WIP) address comments

* (WIP) respond to PR comments

* test refactor

* (wip) consolidating params

* rebase upstream

* import params fix

* (wip) TFInputGraph impl

* (wip) moving to new API

* (wip) enable saved_model tests

* (wip) enable checkpoint test

* (wip) enable multiple tensor tests

* enable all tests

* optimize graph for inference

* allows setting TFInputGraph

* utilize test_input_graph for transformer tests

* enable all tests

Signed-off-by: Philip Yang <philip.yang@databricks.com>

* input graph

* docs

* tensor tests

* tensor test update

* TFTransformer Part-4 Test Refactor (#15)

* adding new tests

* remove original test design

* cleanup

* deleting original testing ideas

* PR comments
